LochNessBuilder/Builder.cs

```diff
@@ -142,8 +142,18 @@ public Builder<TInstance> WithFactory<TProp>(Expression<Func<TInstance, TProp>>
             var instance = GetInstance();
             var prop = GetProp(selector, instance);
             Expression<Func<TProp>> valueInvoker = () => value();
+            Expression setExpression;
+
+            var unaryExpression = selector.Body as UnaryExpression;
+            if (unaryExpression != null && unaryExpression.NodeType == ExpressionType.Convert)
```

Comment:
I'd be tempted to add some comments showing what case we're supporting with what line. Are we supporting soemthing like:
`.WithFactory(x => !x.Party, () => false)` ? Why is this useful? Or more like:
 // x.Count is long
`.WithFactory(x => x.Count, 0) // 0 is int by default, x.Count is long. The expression is now a Unary Convert > Assign`

Also - how does the expression in line 150 do setting? I'm intrigued

EG I'd imagine we'd need something like  `Expression.Assign(prop, Expression.Invoke(Expression.Convert(valueInvoker.Body, typeof(TProp))))` but I imagine it's just my poor understanding of what's going on!

Also - are there any other type of Expressions we don't support? I imagine it's either UnaryExpression or MemberExpression - else we should error? (So potentially we could use a switch statement - or even c# 8 switch statement)

If we instead specify that the method has to take a `MemberExpression`, is the compiler super clever and will realise TProp has to be eg the long in this case, and then insert an implict case into the assignment lambda which feels more "correct"?

Reply:
Thoughts on this as a Doc block:

```
        /* ========================
         * ==  Unary Conversions ==
         * ========================
         *
         * Most calls looks something like this:
         *     `.With(obj => obj.IntProp, 2)`
         *
         * But some calls might look like this:
         *     `.With(obj => obj.ShortProp, 2019)`
         * where the prop type and the argument are not the same, but have implicit casts available.
         *
         * Unfortunately, then compiler turns this into:
         *     `.With<int>(obj => (int)(obj.ShortProp), 2019)`
         * i.e. rather than cast the argument, it assumes an implicit cast on the selector expression.
         * This is the scenario that we're trying to handle with these two methods.
         * One to extract the `obj.ShortProp` out in either case, and one to explicitly detect the casting case
         * as we need to know that the `TProp` that we think we have isn't actually correct.
         * (i.e. `obj.ShortProp` is a `short`, not the `int` that we have in `TProp`.)
         *
         * In most *realistic* situations, we will be fine to convert the value, rather than the Property.
         * Bu it is entirely possible to write code that compiles, but which can't execute:
         *     `.With(obj => obj.ChildTypeProp, new ParentType())`
         * As in the obj.ShortProp example, this will compile, as it can be interpretted as:
         *     `.With<ParentType>(obj => (ParentType)(obj.ChildTypeProp), new ParentType())`
         * but it cannot possibly succeed, as it is not possible to assign a `new ParentType()` to `obj.ChildTypeProp`.
         * The relevant ArgumentException will be thrown, on the `.With()`, and the user should be able to fix the mis-usages.
         */
```

Reply:
As discussed. Doesn't work, does indeed need a setter included. Have modified, and put a "see comment in `.IsUnaryConversion()`" comment reference